piwik-sdk/src/main/java/org/piwik/sdk/CustomVariables.java

```diff
@@ -73,21 +72,25 @@ public JSONArray put(int index, String name, String value) {
      * @param values packed key/value pair
      * @return super.put result or null if key is null or value length is not equals 2
      */
-    @Override
     public JSONArray put(String index, JSONArray values) {
         if (values.length() == 2 && index != null) {
-            return super.put(index, values);
+            synchronized (mConcurrentLock) {
+                return mVars.put(index, values);
+            }
         }
         Logy.d(LOGGER_TAG, "value length should be equal 2");
         return null;
     }
 
-    @Override
     public String toString() {
-        if (size() == 0) {
+        if (mVars.size() == 0) {
```

Comment:
Assuming the change to ConcurrentHashMap.
```
        JSONObject json = new JSONObject(mVars);
        return json.length() > 0 ? json.toString() : null;
```

There is also a race condition issue this fixes (we have not noticed yet).
After passing the `== 0` check, the map can still be emptied before we copy it.
So then we would return an empty String instead of a `null` String.